DataFormats/Phase2TrackerDigi/src/Phase2ITQCore.cc

```diff
@@ -0,0 +1,167 @@
+#include "DataFormats/Phase2TrackerDigi/interface/Phase2ITQCore.h"
+#include <cmath>
+#include <vector>
+#include <iostream>
+
+//4x4 region of hits in sensor coordinates
+Phase2ITQCore::Phase2ITQCore(int rocid,
+                             int ccol_in,
+                             int qcrow_in,
+                             bool isneighbour_in,
+                             bool islast_in,
+                             std::vector<int> adcs_in,
+                             std::vector<int> hits_in) {
+  rocid_ = rocid;
+  ccol = ccol_in;
+  qcrow = qcrow_in;
+  isneighbour_ = isneighbour_in;
+  islast_ = islast_in;
+  adcs = adcs_in;
+  hits = hits_in;
+}
+
+//Takes a hitmap in sensor coordinates in 4x4 and converts it to readout chip coordinates with 2x8
+std::vector<bool> Phase2ITQCore::toRocCoordinates(std::vector<bool>& hitmap) {
+  std::vector<bool> ROC_hitmap(16, 0);
+
+  for (size_t i = 0; i < hitmap.size(); i++) {
+    int row = std::floor(i / 4);
+    int col = i % 4;
+    int new_row;
+    int new_col;
+
+    if (row % 2 == 0) {
+      new_row = row / 2;
+      new_col = 2 * col;
+    } else {
+      new_row = std::floor(row / 2);
+      new_col = 2 * col + 1;
+    }
+
+    int new_index = 8 * new_row + new_col;
+    ROC_hitmap[new_index] = hitmap[i];
+  }
+
+  return ROC_hitmap;
+}
+
+//Returns the hitmap for the Phase2ITQCore in 4x4 sensor coordinates
+std::vector<bool> Phase2ITQCore::getHitmap() {
+  std::vector<bool> hitmap = {};
+
+  for (auto hit : hits) {
+    hitmap.push_back(hit > 0);
+  }
+
+  return (toRocCoordinates(hitmap));
+}
+
+std::vector<int> Phase2ITQCore::getADCs() {
+  std::vector<int> adcmap = {};
+
+  for (auto adc : adcs) {
+    adcmap.push_back(adc);
+  }
+
+  return adcmap;
+}
+
+//Converts an integer into a binary, and formats it with the given length
+std::vector<bool> Phase2ITQCore::intToBinary(int num, int length) {
```

Comment:
I have to do the block below, in `encodeQCore` function since all these have to be pushed into `std::vector<bool> code = {};`

```
    if (is_new_col) {
        std::bitset<16> col_code = intToBinary(ccol, 6);
        for (int i = 0; i < 6; ++i) {
            code.push_back(col_code[6 - 1 - i]);  // Push only 6 bits in correct order
        }
    }
```

instead of doing 

```
  if (is_new_col) {               
    std::vector<bool> col_code = intToBinary(ccol, 6);
    code.insert(code.end(), col_code.begin(), col_code.end());
  }                               
```

because `std::bitset<16>` gives back fixed number of bits everytime which changes in different lines
`    std::vector<bool> col_code = intToBinary(ccol, 6);`
`    std::vector<bool> row_code = intToBinary(qcrow, 8);`
`    std::vector<bool> adc_code = intToBinary(adc, 4);`
and i only want 6, 8, 4 bits to be pushed.

is it still preferred? (or any other suggestions will be grateful)

Reply:
So is the maximum length of the final bitfield known only at runtime when `Phase2ITQCore::encodeQCore()` is called? (I'd guess that because `Phase2ITQCore` constructor takes the contents of `acds_` data member from argument)

If that is the case, using `std::vector<bool>` (or similar) in the end sounds unavoidable. Then I'm just thinking out loud if code along
```cpp
  auto appendToCode = [&code](auto bits) {
    for (int i = 0; i < bits.size(); ++i)
      code.push_back(bits[i]);
  };
  if (is_new_col) {
    appendToCode(std::bitset<6>(ccol_));
  }
  // ...
  if (!isneighbour_) {
    appendToCode(std::bitset<8>(qcrow_));
  }
  // ...
```
would be more economical (at least it would avoid the memory allocations for the intermediate `std::vector<bool>`s, and the need for a custom `intToBinary`).

Or wait, does the `intToBinary` turn the bits around? I.e. the most significant bit of `num` (as specified by `length`) becomes the first element (i.e. index `0`) of the `bi_num`? In that case the aforementioned `appendToCode` would be
```cpp
  auto appendToCode = [&code](auto bits) {
    for (int i = bits.size()-1; i >= 0; --i)
      code.push_back(bits[i]);
  };
```